// @nrwl/nx-cloud/lib/nx-cloud-tasks-runner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const rxjs_1 = require("rxjs");
const tasks_runner_v2_1 = require("@nrwl/workspace/src/tasks-runner/tasks-runner-v2");
const output_1 = require("@nrwl/workspace/src/utils/output");
const app_root_1 = require("@nrwl/workspace/src/utils/app-root");
const fs = require("fs");
const path = require("path");
const uuid_1 = require("uuid");
const crypto = require("crypto");
const node_machine_id_1 = require("node-machine-id");
const os = require("os");
const path_1 = require("path");
const axios = require('axios');
const tar = require('tar');
const chalk = require('chalk');
const INVOKED_BY_RUNNER = process.env.NX_INVOKED_BY_RUNNER;
class CloudApi {
    constructor(messages, runContext, baseUrl, authToken, machineInfo) {
        this.messages = messages;
        this.runContext = runContext;
        this.machineInfo = machineInfo;
        this.apiAxiosInstance = axios.create({
            baseURL: baseUrl,
            timeout: 10000,
            headers: { authorization: authToken },
        });
    }
    startRun(hashes) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const resp = yield this.apiAxiosInstance.post('/nx-cloud/runs/start', {
                    meta: {
                        nxCloudVersion: this.nxCloudVersion(),
                    },
                    hashes,
                });
                if (resp.data && resp.data.message) {
                    this.messages.message = resp.data.message;
                }
                if (!resp.data || !resp.data.urls) {
                    this.messages.apiError = `Invalid Nx Cloud response: ${JSON.stringify(resp.data)}`;
                    return {};
                }
                return resp.data.urls;
            }
            catch (e) {
                this.messages.apiError = this.messages.extractErrorMessage(e, 'api');
                return {};
            }
        });
    }
    endRun(run, tasks) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                // API is not working, don't make the end request
                if (this.messages.apiError)
                    return;
                const resp = yield this.apiAxiosInstance.post('/nx-cloud/runs/end', {
                    meta: {
                        nxCloudVersion: this.nxCloudVersion(),
                    },
                    tasks,
                    run: run,
                    machineInfo: this.machineInfo,
                });
                if (resp.data && resp.data.status) {
                    if (resp.data.status !== 'success') {
                        this.messages.apiError = `Invalid end run response: ${JSON.stringify(resp.data.message)}`;
                    }
                    else {
                        if (resp.data.runUrl) {
                            this.runContext.runUrl = resp.data.runUrl;
                        }
                    }
                }
                else if (resp.data && typeof resp.data === 'string') {
                    if (resp.data !== 'success') {
                        this.messages.apiError = `Invalid end run response: ${JSON.stringify(resp.data)}`;
                    }
                }
                else {
                    this.messages.apiError = `Invalid end run response: ${JSON.stringify(resp.data)}`;
                }
            }
            catch (e) {
                this.messages.apiError = this.messages.extractErrorMessage(e, 'api');
            }
        });
    }
    nxCloudVersion() {
        try {
            const v = JSON.parse(fs.readFileSync(`package.json`).toString());
            return v.devDependencies['@nrwl/nx-cloud'];
        }
        catch (e) {
            return 'unknown';
        }
    }
    findSimilarTasks(task) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                // API is not working, don't make the end request
                if (this.messages.apiError)
                    return null;
                const resp = yield this.apiAxiosInstance.post('/nx-cloud/find-similar-tasks', {
                    task: {
                        target: task.target,
                        projectName: task.projectName,
                        hash: task.hash,
                        hashDetails: task.hashDetails,
                    },
                });
                if (!resp.data) {
                    this.messages.apiError = `Invalid end run response: ${JSON.stringify(resp.data)}`;
                }
                return resp.data.message;
            }
            catch (e) {
                this.messages.apiError = this.messages.extractErrorMessage(e, 'api');
                return null;
            }
        });
    }
}
class E2EEncryption {
    constructor(key) {
        if (key) {
            this.encryptionKey = this.to32bytes(key);
        }
    }
    to32bytes(key) {
        let res = key;
        while (res.length < 32) {
            res += key;
        }
        return Buffer.from(res).slice(0, 32);
    }
    hasEncryption() {
        return !!this.encryptionKey;
    }
    encryptFile(file) {
        const iv = crypto.randomBytes(16);
        const cipher = crypto.createCipheriv('aes-256-cbc', this.encryptionKey, iv);
        const decryptedFileContents = fs.readFileSync(file);
        const e = cipher.update(decryptedFileContents);
        const encryptedFileContents = Buffer.concat([iv, e, cipher.final()]);
        fs.writeFileSync(file, encryptedFileContents);
    }
    decryptFile(file) {
        const encryptedFileContents = fs.readFileSync(file);
        try {
            const decipher = crypto.createDecipheriv('aes-256-cbc', this.encryptionKey, encryptedFileContents.slice(0, 16) // iv
            );
            const encryptedText = encryptedFileContents.slice(16); // remove the iv
            const d = decipher.update(encryptedText);
            const decryptedFileContents = Buffer.concat([d, decipher.final()]);
            fs.writeFileSync(file, decryptedFileContents);
        }
        catch (e) {
            throw new Error(`Could not decrypt the artifact. Please check your encryption key.`);
        }
    }
}
class CloudRemoteCache {
    constructor(messages, urls, runContext, encryption) {
        this.messages = messages;
        this.urls = urls;
        this.runContext = runContext;
        this.encryption = encryption;
        this.storeRequests = [];
    }
    retrieve(hash, cacheDirectory) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.messages.cacheError)
                return false;
            try {
                const hashUrls = (yield this.urls)[hash];
                if (!hashUrls || !hashUrls.get) {
                    this.runContext.statuses[hash] = 'remote-cache-miss';
                    return false;
                }
                if (CloudRemoteCache.VERBOSE_LOGGING) {
                    output_1.output.note({ title: `Nx Cloud: Downloading ${hash}` });
                }
                const tgz = this.createFileName(hash, cacheDirectory);
                yield this.downloadFile(hashUrls.get, tgz);
                this.createCommitFile(hash, cacheDirectory);
                if (CloudRemoteCache.VERBOSE_LOGGING) {
                    output_1.output.note({ title: `Nx Cloud: Cache hit ${hash}` });
                }
                this.runContext.statuses[hash] = 'remote-cache-hit';
                return true;
            }
            catch (e) {
                if (e.response && e.response.status === 404) {
                    if (CloudRemoteCache.VERBOSE_LOGGING) {
                        output_1.output.note({ title: `Nx Cloud: Cache miss ${hash}` });
                    }
                    // cache miss. print nothing
                }
                else {
                    this.messages.cacheError = this.messages.extractErrorMessage(e, 'storage');
                }
                this.runContext.statuses[hash] = 'remote-cache-miss';
                return false;
            }
        });
    }
    store(hash, cacheDirectory) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.messages.cacheError)
                return false;
            const res = Promise.resolve().then(() => __awaiter(this, void 0, void 0, function* () {
                const hashUrls = (yield this.urls)[hash];
                if (!hashUrls)
                    return false;
                if (!hashUrls.put) {
                    if (CloudRemoteCache.VERBOSE_LOGGING) {
                        output_1.output.note({
                            title: `Nx Cloud: Skipping storing ${hash}. Read only token is used.`,
                        });
                    }
                    return true;
                }
                try {
                    if (CloudRemoteCache.VERBOSE_LOGGING) {
                        output_1.output.note({ title: `Nx Cloud: Storing ${hash}` });
                    }
                    const tgz = yield this.createFile(hash, cacheDirectory);
                    yield this.uploadFile(hashUrls.put, tgz);
                    if (CloudRemoteCache.VERBOSE_LOGGING) {
                        output_1.output.note({ title: `Nx Cloud: Stored ${hash}` });
                    }
                    return true;
                }
                catch (e) {
                    this.messages.cacheError = this.messages.extractErrorMessage(e, 'storage');
                    return false;
                }
            }));
            this.storeRequests.push(res);
            return res;
        });
    }
    waitForStoreRequestsToComplete() {
        return Promise.all(this.storeRequests);
    }
    createFileName(hash, cacheDirectory) {
        return path.join(cacheDirectory, `${hash}.tar.gz`);
    }
    downloadFile(url, tgz) {
        return __awaiter(this, void 0, void 0, function* () {
            const resp = yield axios(url, {
                method: 'GET',
                responseType: 'stream',
                maxContentLength: 1000 * 1000 * 200,
                timeout: 30000,
            });
            if (this.encryption.hasEncryption()) {
                yield new Promise((res) => {
                    const f = resp.data.pipe(fs.createWriteStream(tgz));
                    f.on('close', () => res());
                });
                this.encryption.decryptFile(tgz);
                const q = fs.createReadStream(tgz).pipe(tar.x({
                    cwd: path.dirname(tgz),
                }));
                return new Promise((res) => {
                    q.on('close', () => {
                        res();
                    });
                });
            }
            else {
                const q = resp.data.pipe(tar.x({
                    cwd: path.dirname(tgz),
                }));
                return new Promise((res) => {
                    q.on('close', () => {
                        res();
                    });
                });
            }
        });
    }
    createCommitFile(hash, cacheDirectory) {
        fs.writeFileSync(path.join(cacheDirectory, `${hash}.commit`), 'true');
    }
    createFile(hash, cacheDirectory) {
        return __awaiter(this, void 0, void 0, function* () {
            const tgz = this.createFileName(hash, cacheDirectory);
            yield tar.c({
                gzip: true,
                file: tgz,
                cwd: cacheDirectory,
            }, [hash]);
            if (this.encryption.hasEncryption()) {
                this.encryption.encryptFile(tgz);
            }
            return tgz;
        });
    }
    uploadFile(url, tgz) {
        return __awaiter(this, void 0, void 0, function* () {
            yield axios(url, {
                method: 'PUT',
                data: fs.readFileSync(tgz),
                headers: { 'Content-Type': 'application/octet-stream' },
                maxContentLength: 1000 * 1000 * 200,
            });
        });
    }
}
CloudRemoteCache.VERBOSE_LOGGING = process.env.NX_VERBOSE_LOGGING;
class TaskRunnerLifeCycle {
    constructor(runContext, cacheDirectory, collectTerminalOutput) {
        this.runContext = runContext;
        this.cacheDirectory = cacheDirectory;
        this.collectTerminalOutput = collectTerminalOutput;
        this.tasks = [];
        this.uniqRunId = uuid_1.v4();
    }
    startTask(task) {
        const overrides = Object.assign({}, (task.overrides || {}));
        if (task.target.configuration) {
            overrides['configuration'] = task.target.configuration;
        }
        const params = Object.keys(overrides).reduce((m, k) => `${m} --${k}=${overrides[k]}`, ``);
        this.tasks.push({
            taskId: `${task.id}-${this.uniqRunId}`,
            startTime: new Date().toISOString(),
            target: task.target.target,
            projectName: task.target.project,
            hash: task.hash,
            hashDetails: task.hashDetails,
            params: params,
        });
    }
    endTask(task, code) {
        const startedTask = this.tasks.find((t) => t.taskId === `${task.id}-${this.uniqRunId}`);
        if (!startedTask) {
            throw new Error(`Cannot find task ${task.id}`);
        }
        startedTask.endTime = new Date().toISOString();
        startedTask.status = code;
        startedTask.cacheStatus =
            this.runContext.statuses[task.hash] || 'no-remote-cache';
        if (this.collectTerminalOutput && !INVOKED_BY_RUNNER) {
            startedTask.terminalOutput = this.getTerminalOutput(task);
        }
    }
    getTerminalOutput(task) {
        let dir;
        if (this.cacheDirectory) {
            if (this.cacheDirectory.startsWith('./')) {
                dir = path_1.join(app_root_1.appRootPath, this.cacheDirectory);
            }
            else {
                dir = this.cacheDirectory;
            }
        }
        else {
            dir = path_1.join(app_root_1.appRootPath, 'node_modules', '.cache', 'nx');
        }
        try {
            const taskOutput = this.readTerminalOutputFile(dir, task.hash);
            return taskOutput.length > 50000
                ? `${taskOutput.slice(0, 50000)}...\nTRUNCATED`
                : taskOutput;
        }
        catch (e) {
            return '';
        }
    }
    readTerminalOutputFile(dir, hash) {
        try {
            return fs
                .readFileSync(path.join(dir, 'terminalOutputs', hash))
                .toString();
        }
        catch (_a) {
            try {
                return fs
                    .readFileSync(path.join(dir, hash, 'terminalOutput'))
                    .toString();
            }
            catch (_b) {
                return '';
            }
        }
    }
}
class MessageReporter {
    constructor(options) {
        this.options = options;
        this.cacheError = null;
        this.apiError = null;
        this.message = null;
    }
    get anyErrors() {
        return this.cacheError || this.apiError;
    }
    printMessages() {
        if (this.anyErrors) {
            const bodyLines = [];
            if (this.cacheError) {
                bodyLines.push(`- ${this.cacheError}`);
            }
            if (this.apiError && this.apiError !== this.cacheError) {
                bodyLines.push(`- ${this.apiError}`);
            }
            output_1.output.warn({
                title: `Nx Cloud Problems`,
                bodyLines,
            });
        }
        if (this.message) {
            printMessage(this.message);
        }
    }
    extractErrorMessage(e, scope) {
        if (e.code === 'ECONNREFUSED' ||
            e.code === 'EAI_AGAIN' ||
            e.code === 'ENOTFOUND' ||
            e.code === 'EPROTO' ||
            e.code === 'ECONNABORTED') {
            return `Cannot connect to remote cache (scope: ${scope}, code: ${e.code}).`;
        }
        else if (e.response && e.response.status === 401) {
            return e.response.data.message
                ? e.response.data.message
                : e.response.data;
        }
        else if (e.response && e.response.status === 402) {
            if (this.options.showUsageWarnings === false)
                return null;
            return e.response.data.message
                ? e.response.data.message
                : e.response.data;
        }
        else {
            let details = '';
            if (e.response && e.response.data && e.response.data.message) {
                details = `. ${e.response.data.message}`;
            }
            else if (e.response && e.response.data) {
                details = `. ${e.response.data}`;
            }
            return `${e.message}${details} (code: ${e.code})`;
        }
    }
}
class EndOfRunMessage {
    constructor(messages, runContext) {
        this.messages = messages;
        this.runContext = runContext;
        this.tasks = [];
        this.patchReporters();
    }
    // We patch reporters temporary. Nx 10 should have a proper way of printing remote cache hits.
    patchReporters() {
        try {
            const reporter = require('@nrwl/workspace/src/tasks-runner/default-reporter');
            const old = reporter.DefaultReporter.prototype.printResults;
            const that = this;
            reporter.DefaultReporter.prototype.printResults = function () {
                old.apply(this, arguments);
                that.printCacheHitsMessage();
            };
        }
        catch (e) { }
        try {
            const reporter = require('@nrwl/workspace/src/tasks-runner/run-one-reporter');
            const old = reporter.RunOneReporter.prototype.printResults;
            const that = this;
            reporter.RunOneReporter.prototype.printResults = function () {
                old.apply(this, arguments);
                that.printCacheHitsMessage();
            };
        }
        catch (e) { }
        try {
            const reporter = require('@nrwl/workspace/src/tasks-runner/empty-reporter');
            const old = reporter.EmptyReporter.prototype.printResults;
            const that = this;
            reporter.EmptyReporter.prototype.printResults = function () {
                old.apply(this, arguments);
                that.printCacheHitsMessage();
            };
        }
        catch (e) { }
    }
    printCacheHitsMessage() {
        if (this.messages.anyErrors)
            return;
        if (INVOKED_BY_RUNNER)
            return;
        const hits = this.tasks
            .filter((t) => this.runContext.statuses[t.hash] === 'remote-cache-hit')
            .map((t) => t.projectName);
        if (hits.length === 0 && !this.runContext.runUrl) {
            return;
        }
        const projects = hits.length === 1 ? hits[0] : `${hits.length} projects`;
        const message = [];
        if (hits.length > 0) {
            message.push(`Results for ${projects} were retrieved from Nx Cloud.`);
        }
        if (this.runContext.runUrl) {
            message.push(`See run details at ${this.runContext.runUrl}`);
        }
        if (message.length > 0) {
            printMessage(message.join(' '));
        }
    }
}
function printMessage(message) {
    const formattedMessage = output_1.output.colors.gray(message);
    output_1.output.addNewline();
    output_1.output.writeOutputTitle({
        label: chalk.reset.inverse.bold.keyword('grey')(' CLOUD '),
        title: chalk.keyword('grey')(formattedMessage),
    });
    output_1.output.addNewline();
}
function getMachineInfo(options) {
    let machineId = '';
    if (options.canTrackAnalytics) {
        try {
            machineId = node_machine_id_1.machineIdSync();
        }
        catch (_a) { }
    }
    return {
        machineId,
        platform: os.platform(),
        version: os.version ? os.version() : '',
        cpuCores: os.cpus().length,
    };
}
function createApi(errors, options, runContext) {
    const baseUrl = options.url || 'https://api.nrwl.io';
    const machineInfo = getMachineInfo(options);
    if (process.env.NX_CLOUD_AUTH_TOKEN) {
        return new CloudApi(errors, runContext, baseUrl, process.env.NX_CLOUD_AUTH_TOKEN, machineInfo);
    }
    else {
        return new CloudApi(errors, runContext, baseUrl, options.accessToken, machineInfo);
    }
}
function parseCommand() {
    return `nx ${process.argv.slice(2).join(' ')}`;
}
function runGroup() {
    return process.env.NX_RUN_GROUP;
}
function branch() {
    return process.env.NX_BRANCH;
}
const nxCloudTasksRunner = (tasks, options, context) => {
    if (process.env.NX_CLOUD_AUTH_TOKEN || options.accessToken) {
        const runContext = { statuses: {} };
        const messages = new MessageReporter(options);
        const api = createApi(messages, options, runContext);
        const endOfRunMessage = new EndOfRunMessage(messages, runContext);
        const urls = api.startRun(tasks.map((t) => t.hash));
        const runStartTime = new Date().toISOString();
        const lifeCycle = new TaskRunnerLifeCycle(runContext, options.cacheDirectory, !!options.scan);
        const encryption = new E2EEncryption(process.env.NX_CLOUD_ENCRYPTION_KEY || options.encryptionKey);
        const remoteCache = new CloudRemoteCache(messages, urls, runContext, encryption);
        const res = tasks_runner_v2_1.tasksRunnerV2(tasks, Object.assign(Object.assign({}, options), { remoteCache, lifeCycle }), context);
        const wrappedRes = new rxjs_1.Subject();
        res.subscribe({
            next: (value) => wrappedRes.next(value),
            error: (err) => wrappedRes.error(err),
            complete: () => __awaiter(void 0, void 0, void 0, function* () {
                const runEndTime = new Date().toISOString();
                yield Promise.all([
                    remoteCache.waitForStoreRequestsToComplete(),
                    api.endRun({
                        command: parseCommand(),
                        startTime: runStartTime,
                        endTime: runEndTime,
                        scan: !!options.scan,
                        branch: branch(),
                        runGroup: runGroup(),
                    }, lifeCycle.tasks),
                ]);
                messages.printMessages();
                endOfRunMessage.tasks = lifeCycle.tasks;
                wrappedRes.complete();
            }),
        });
        return wrappedRes;
    }
    else {
        return tasks_runner_v2_1.tasksRunnerV2(tasks, options, context);
    }
};
exports.default = nxCloudTasksRunner;
